### ISSUE TITLE
Fully-configurable OpenType features

### ISSUE THREAD

**Issue by be5invis** [EdiWang/Character-Map-UWP#112]
The current variant selector only allows selecting one variant:
![image](https://user-images.githubusercontent.com/240091/80772278-04398100-8b0b-11ea-9e92-3cb7d70a00c8.png)

However many fonts allows:
 * Applying multiple features at once;
 * Have features that this list currently not covered (like `cv01`);
 * Leverages GSUB's alternative substitutions, so the user needs to pass an integer parameter to the feature (like `aalt=1`).

The enhancement might be changing this dropdown menu into a textbox and let user to input the feature tags they need, like `kern, calt, cv03, cv05, aalt=2`.

**Comment by JohnnyWestlake** [comment]
Internally the app actually does list all the not-covered variations, but main problem is the UI layer (UWP XAML) doesn't support them - it only supports a limited set of pre-defined features that do not allow configuration, and don't behave properly when applied together.

Like a lot of these tied to the UI layer, the solution is to either wait for Microsoft to release WinUI 3 (the open source variant of XAML, so that we can actually fix this properly at the UI layer), or re-write the character map grid entirely in DirectX (which is not impossible but leads to *lot* of extra work around input and keyboards and GPU handling and accessibility).

At the moment I'm not likely to attempt the latter (though anyone can feel free), and WinUI3 is at least 6 months away.

**Comment by be5invis** [comment]
I do not think that regular text controls will support advanced typography that configurable — we should definitely move to custom rendering using `IDWriteTextLayout`.

**Comment by be5invis** [comment]
@JohnnyWestlake I think a good solution is to rewrite only the rendering part of the glyph shape and leave the rest of the control as-is.
![image](https://user-images.githubusercontent.com/240091/80797054-a41bfc80-8b55-11ea-8058-17e84806c4d6.png)

**Comment by JohnnyWestlake** [comment]
(Support for variable non-XAML supported OpenType tags for **_Type Ramp view only_** is currently in the [Character-Map-UWP/Variations-Basic](https://github.com/character-map-uwp/Character-Map-UWP/tree/Variations-Basic) dev branch - this does not include anything with an integer parameter other than 1 for now)

**Comment by be5invis** [comment]
@JohnnyWestlake Non-1 parameters are not frequently used (though some features like `salt` may use that). You may also need to turn _off_ certain default features, like `kern` or `calt`.

**Comment by be5invis** [comment]
@JohnnyWestlake And also, OpenType supports **enabling multiple feature simultaneously**, like both `cv01` and `cv02`.

**Comment by be5invis** [comment]
Also, certain features used for shaping should be hidden. Enabling/Disabling these features are controlled by the shaping engine, not the user. The list includes:

 * `locl`, `ccmp`, `nukt`, `akhn`, `rvrn`
 * `rphf`, `pref`
 * `rkrf`, `abvf`, `blwf`, `half`, `pstf`, `vatu`, `cjct`
 * `isol`, `init`, `medi`, `fina`, `stch`
 * `abvs`, `blws`, `haln`, `pres`, `psts`, `rclt`, `rlig`
 * `curs`, `dist`, `mark`, `abvm`, `blwm`, `mkmk`, `mset`

The following features are considered typographic, but should be enabled by default:

* `calt`, `liga`, `clig`, `kern`
